test/extended/storage/mutable_csinode_allocatable.go

```diff
@@ -0,0 +1,678 @@
+			return false
+		}, resourceExhaustionTimeout, csiNodeUpdatePollInterval).Should(o.BeTrue(),
+			"ResourceExhausted error should be detected")
+
+		// Verify that CSINode allocatable count was updated immediately (not waiting for the long update period)
+		// The count should remain at original-1 because we still have 1 ENI attached
+		g.By(fmt.Sprintf("# Verifying that CSINode allocatable count updated to %d (immediate update triggered by ResourceExhausted)", originalAllocatableCount-1))
+		expectedCount := originalAllocatableCount - 1
+		o.Eventually(func() int32 {
+			currentCount := getCSINodeAllocatableCountByDriver(ctx, oc, targetWorkerNode, ebsCSIDriverName)
+			e2e.Logf("Current CSINode allocatable count: %d, expected: %d", currentCount, expectedCount)
+			return currentCount
+		}, csiNodeUpdatePollTimeout, csiNodeUpdatePollInterval).Should(o.Equal(expectedCount),
+			"CSINode allocatable count should be immediately updated despite long update period")
+		e2e.Logf("Successfully verified that ResourceExhausted error triggered immediate CSINode update")
+
+		g.By("# Verifying the last pod should FailedScheduling after CSINode allocatable count updated")
+		o.Eventually(func() bool {
+			pod, err := oc.AdminKubeClient().CoreV1().Pods(oc.Namespace()).Get(ctx, lastPodName, metav1.GetOptions{})
+			if err != nil {
+				e2e.Logf("Pod %s not found yet: %v", lastPodName, err)
+				return false
+			}
+
+			if pod.Status.Phase != v1.PodPending {
+				e2e.Logf("Pod %s not Pending yet, current phase: %s", lastPodName, pod.Status.Phase)
+				return false
+			}
+
+			events, err := oc.AdminKubeClient().CoreV1().Events(oc.Namespace()).List(ctx, metav1.ListOptions{
+				FieldSelector: fmt.Sprintf("involvedObject.name=%s", lastPodName),
+			})
+			if err != nil {
+				return false
+			}
+
+			for _, e := range events.Items {
+				if e.Reason == "FailedScheduling" && strings.Contains(e.Message, "exceed max volume count") {
+					e2e.Logf("Pod %s is Pending due to volume attach limit as expected", lastPodName)
+					return true
+				}
+			}
+
+			return false
+		}, csiNodeUpdatePollTimeout, csiNodeUpdatePollInterval).Should(o.BeTrue(),
+			"Pod should be Pending due to volume attach limit")
+	})
+
+})
+
+// createStatefulSetWithVolumeTemplate creates a StatefulSet with volumeClaimTemplates
+func createStatefulSetWithVolumeTemplate(namespace, name string, replicas int32, nodeName, storageClassName string) *appsv1.StatefulSet {
+	allowPrivEsc := false
+	return &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: appsv1.StatefulSetSpec{
+			ServiceName: name,
+			Replicas:    &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": name,
+				},
+			},
+			PersistentVolumeClaimRetentionPolicy: &appsv1.StatefulSetPersistentVolumeClaimRetentionPolicy{
+				WhenDeleted: appsv1.DeletePersistentVolumeClaimRetentionPolicyType,
```

Comment:
I would set it to delete the PVCs and then you don't need to delete them in the test.

Reply:
Yeah, it is already set to delete -> https://github.com/openshift/origin/blob/main/vendor/k8s.io/api/apps/v1/types.go#L149